core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -250,12 +266,31 @@ private void advanceToNextPage() {
 
     @Override
     public boolean hasNext() {
-      return currentRecordNumber != numRecords;
+      boolean ret = currentRecordNumber != numRecords;
+      if (destructive && !ret) {
+        // Remove latest page.
+        dataPagesIterator.remove();
+        this.map.taskMemoryManager.freePage(currentPage);
+        this.map.shuffleMemoryManager.release(currentPage.size());
+        this.map.bitset = null;
+        this.map.longArray = null;
+        assert(this.map.dataPages.isEmpty());
+      }
+      return ret;
     }
 
     @Override
     public Location next() {
       int totalLength = PlatformDependent.UNSAFE.getInt(pageBaseObject, offsetInPage);
+      if (destructive) {
+        MemoryLocation keyAddress = loc.getKeyAddress();
+        Object keyBaseObject = keyAddress.getBaseObject();
+        long keyBaseOffset = keyAddress.getBaseOffset();
+
+        int hashcode = HASHER.hashUnsafeWords(keyBaseObject, keyBaseOffset, loc.getKeyLength());
+        int pos = hashcode & map.mask;
+        this.map.bitset.unset(pos);
+      }
       if (totalLength == END_OF_PAGE_MARKER) {
```

Comment:
I did that in `advanceToNextPage` above, do I need to move it here? I think it is no difference.